### PR TITLE
Fix author information leakage by author blocks for Custom Post Types without author support & display notice to user

### LINKED ISSUE #65816: Author blocks display author information for CPT that does not support authors
### Description

The various author information blocks will display the author information even when the post belongs to a Custom Post Type that does not declare `author` support. Because WP will still save an author into the database, this essentially leaks information that's not intended to be visible.

Sent here from https://core.trac.wordpress.org/ticket/62152

### Step-by-step reproduction instructions

1. Create a custom post type and specify that it does not support `author`
2. Create a new post of that post type
3. View that post on the front end
4. The author is shown on the page, even though the post type does not support `author`

#### Screenshot (Twenty Twenty-Four)

Screenshot 2024-10-02 at 09 45 17

#### CPT declaration

```php
add_action( 'init', function() {
	register_post_type( 'custom-post', array(
		'public' => true,
		'supports' => array(
			0 => 'title',
			1 => 'editor',
			2 => 'custom-fields',
			/* No support for 'author' declared */
		),
		'delete_with_user' => false,
	) );
} );
```

### Environment info

- WordPress 6.6.2
- Bundled themes Twenty Twenty-Four, Twenty Twenty-Three, or Twenty Twenty-Two

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [X] Yes

### PR DESCRIPTION
**Pull request by sarthaknagoshe2002**
fixes: #65816
## What?
This PR updates the behavior of author blocks to ensure they do not render when the post belongs to a Custom Post Type (CPT) that does not explicitly declare support for the author feature.

## Why?
Author blocks displaying on CPTs without declared author support is inconsistent with WordPress behavior.
This can lead to unintentional data exposure, as WordPress still saves author information to the database even when it is not intended to be visible.
Ensures better security, consistency, and adherence to user expectations for CPTs.

## How?

- Added a conditional check in the rendering logic of the author blocks to verify if the CPT supports the author feature.
- If the feature is unsupported, a message is displayed in the editor notifying users that the block will not render on the frontend.
- Updated the frontend output to prevent leaking author information when the CPT does not support authors.

## Testing Instructions
Reproduction:-
1. Create a Custom Post Type (CPT) without declaring support for the author feature:

```php
add_action( 'init', function() {
    register_post_type( 'custom-post', array(
        'public' => true,
        'supports' => array( 'title', 'editor', 'custom-fields' ),
    ) );
} );
```

2. Add a post to the newly created CPT. 
3. Add an author-related block (e.g., Post Author) in the block editor for that CPT post.
4. View the post on the frontend.

Expected Outcome:-

1. In the Editor:
    - A message is displayed informing users that the author block will not render because the CPT does not support the author feature.

2. On the Frontend:
    - The author block does not render, and no author information is displayed.

## Screenshots or screencast 

https://github.com/user-attachments/assets/0b8bde8b-b032-44bc-b3da-03b04f8cc441

The suggestion given by @dhruvang21 is also used in this PR from the [comment](https://github.com/WordPress/gutenberg/issues/65816#issuecomment-2396080596)

### CODE REVIEW

**Comment by Mamaduka** [comment]
Thanks for contributing, @sarthaknagoshe2002!

I'm unsure if we need to modify the block's edit components. People with access to the editors can also access details about post type author.

If it's decided to display a notice in edit components, it should probably be a part of rendered content and not an editor notice.

cc @WordPress/gutenberg-design

**Comment by sarthaknagoshe2002** [comment]
> If it's decided to display a notice in edit components, it should probably be a part of rendered content and not an editor notice.

@Mamaduka Can you elaborate more on this?

**Comment by Mamaduka** [comment]
The blocks could render static text/warning in the editor when the post type doesn't support authors. But as I mentioned, I'm unsure if that warning makes sense in the editor context.

## Rought Example
CleanShot 2024-11-20 at 19 21 46

**Comment by sarthaknagoshe2002** [comment]
@Mamaduka I have implemented your suggestion now.
In future if we conclude that the warning makes sense in the editor context then we can proceed with the merge.

## Screenshot
Screenshot 2024-11-21 at 12 34 10 AM

**Comment by sarthaknagoshe2002** [comment]
I believe that the warning definitely makes sense in the editor context since displaying warnings/errors on the frontend doesn't seem appropriate.

Also, as you suggested static warning makes sense too.

**Comment by sarthaknagoshe2002** [comment]
@Mamaduka Any updates on how we should proceed with this PR?

File: packages/block-library/src/post-author-name/edit.js [hunk 1]
```diff
@@ -33,6 +33,9 @@ function PostAuthorNameEdit( {
 
 			return {
 				authorName: _authorId ? getUser( _authorId ) : null,
+				supportsAuthor:
+					select( coreStore ).getPostType( postType )?.supports
+						?.author ?? false,
```
Comment (Mamaduka) [inline packages/block-library/src/post-author-name/edit.js]: Let's extract `getPostType` above, alongside the `getUser`.

File: packages/block-library/src/post-author-name/edit.js [hunk 2]
```diff
@@ -90,7 +93,15 @@ function PostAuthorNameEdit( {
...
-			<div { ...blockProps }> { displayAuthor } </div>
+			{ supportsAuthor ? (
+				<div { ...blockProps }> { displayAuthor } </div>
+			) : (
+				<div { ...blockProps }>
+					{ __(
+						'The current post type does not support authors. The Post Author Name block will not be displayed.'
+					) }
+				</div>
+			) }
```
Comment (Mamaduka) [inline packages/block-library/src/post-author-name/edit.js]: I think we can just do condition inside the `<div { ...blockProps }>`.

```js
{ supportsAuthor ? displayAuthor : __( 'The current post type does not support authors. The Post Author Name block will not be displayed.' ) }
```

File: packages/block-library/src/post-author/edit.js [hunk 3]
```diff
@@ -52,6 +52,9 @@ function PostAuthorEdit( {
 				authorId: _authorId,
 				authorDetails: _authorId ? getUser( _authorId ) : null,
 				authors: getUsers( AUTHORS_QUERY ),
+				supportsAuthor:
+					select( coreStore ).getPostType( postType )?.supports
+						?.author ?? false,
```
Comment (Mamaduka) [inline packages/block-library/src/post-author/edit.js]: Let's extract the `getPostType` above alongside the `getUser`.

File: packages/block-library/src/post-author/edit.js [hunk 4]
```diff
@@ -189,55 +192,65 @@ function PostAuthorEdit( {
...
+									__html: authorDetails?.description,
+								} }
+							/>
+						) }
+					</div>
+				</div>
+			) : (
+				<div { ...blockProps }>
+					{ __(
+						'The current post type does not support authors. The Post Author block will not be displayed.'
```
Comment (Mamaduka) [inline packages/block-library/src/post-author/edit.js]: Let's use early return for this instead for condition. It should be more readable.

You could do that just below the `showAuthorControl` variable definition.

**Review by Mamaduka — COMMENTED** [review]
@sarthaknagoshe2002, I left a couple of suggestions. Still looking for design feedback before final approval.

P.S. There's also a proposal to deprecate Post Author block - #53427, but the information leaking issue also applies to deprecated blocks.

**Comment by jameskoster** [comment]
> If it's decided to display a notice in edit components, it should probably be a part of rendered content and not an editor notice.

I agree with this, assuming you mean these inline warning notices (obviously with a different message):

Screenshot 2024-12-17 at 14 22 56

Something like:

> This post type ($post_type) does not support Authors. [Remove block]

Additionally, should those blocks be hidden in the Inserter when editing a template associated with a post type that doesn't support them?

**Comment by Mamaduka** [comment]
Thanks, @jameskoster!

Yes, the block editor warning component can work in this case.

> Additionally, should those blocks be hidden in the Inserter when editing a template associated with a post type that doesn't support them?

Generally, a template-aware inserter would be cool, but hiding blocks without context as to why they're hidden can be confusing.

File: packages/block-library/src/post-author-name/index.php [hunk 5]
```diff
@@ -26,6 +26,10 @@ function render_block_core_post_author_name( $attributes, $content, $block ) {
 		return '';
 	}
 
+	if ( ! post_type_supports( get_post_type( $block->context['postId'] ), 'author' ) ) {
```
Comment (fabiankaegy) [inline packages/block-library/src/post-author-name/index.php]:
```suggestion
	if ( ! post_type_supports( $block->context['postType'], 'author' ) ) {
```

We can directly get the post type via context as long as it's listed in the `usesContext` array in the `block.json`

**Comment by sarthaknagoshe2002** [comment]
@fabiankaegy @Mamaduka I’ve addressed the feedback.  

@jameskoster I’ve updated the message to `This post type ($post_type) does not support Authors`. However, I’m unsure how to implement the `[Remove block]` functionality. Could you please guide me on this?

**Comment by jameskoster** [comment]
I'm not actually sure myself, sorry. Maybe @Mamaduka knows 🤞

File: packages/block-library/src/post-author/edit.js [hunk 6]
```diff
@@ -97,6 +99,18 @@ function PostAuthorEdit( {
 	const showAuthorControl =
 		!! postId && ! isDescendentOfQueryLoop && authorOptions.length > 0;
 
+	if ( ! supportsAuthor ) {
+		return (
+			<div { ...blockProps }>
+				{ sprintf(
+					// Translators: %s is replaced with the name of the post type.
+					__( 'This post type (%s) does not support Authors.' ),
```
Comment (Mamaduka) [inline packages/block-library/src/post-author/edit.js]:
```suggestion
					// translators: %s: Name of the post type e.g: "post".
					__( 'This post type (%s) does not support the author.' ),
```

File: packages/block-library/src/post-author-name/edit.js [hunk 7]
```diff
@@ -90,7 +93,17 @@ function PostAuthorNameEdit( {
...
 			</InspectorControls>
-			<div { ...blockProps }> { displayAuthor } </div>
+			<div { ...blockProps }>
+				{ supportsAuthor
+					? displayAuthor
+					: sprintf(
+							// Translators: %s is replaced with the name of the post type.
+							__(
+								'This post type (%s) does not support Authors.'
+							),
```
Comment (Mamaduka) [inline packages/block-library/src/post-author-name/edit.js]:
```suggestion
							// translators: %s: Name of the post type e.g: "post".
							__(
								'This post type (%s) does not support the author'
							),
```

**Comment by fabiankaegy** [comment]
@sarthaknagoshe2002 you can remove a block like so:

```js
const BlockEdit = (props) => {
    const { clientId } = props;
    const { removeBlock } = useDispatch( blockEditorStore );
    
    return (
        <Button onClick={() => removeBlock(clientId)} isDestructive>Remove</Button>
    );
}
```

**Review by Mamaduka — COMMENTED** [review]
I think there's no need to use the warning component and separate the removal action. Why?

* The severity of missing `author` support is negligible. It won't break or affect anything on the front end. You don't need the Waning component to grab the user's attention.
* Users can use toolbar options to remove the block.

@sarthaknagoshe2002 left minor suggestions regarding translation comments and string. After that, I think we're good to merge.

## Screenshot
CleanShot 2024-12-18 at 18 21 54

**Comment by Mamaduka** [comment]
@fabiankaegy, the block editor component will receive the `onRemove` callback via props. But as I mentioned in my previous [comment](https://github.com/WordPress/gutenberg/pull/67136#pullrequestreview-2511977480), I think there's no need to implement it.

https://github.com/WordPress/gutenberg/blob/3159fa2864b9f1cd560a82cd81550eff1f66616a/packages/block-editor/src/components/block-list/block.js#L126

**Comment by fabiankaegy** [comment]
> @fabiankaegy, the block editor component will receive the `onRemove` callback via props. But as I mentioned in my previous [comment](https://github.com/WordPress/gutenberg/pull/67136#pullrequestreview-2511977480), I think there's no need to implement it.
> 
> 
> 
> https://github.com/WordPress/gutenberg/blob/3159fa2864b9f1cd560a82cd81550eff1f66616a/packages/block-editor/src/components/block-list/block.js#L126

🤯 very good to know thank you :)

**Comment by sarthaknagoshe2002** [comment]
@Mamaduka I’ve addressed the changes, and I believe this is now ready to be merged.

File: packages/block-library/src/post-author-name/edit.js [hunk 8]
```diff
@@ -90,7 +93,17 @@ function PostAuthorNameEdit( {
...
 				</PanelBody>
 			</InspectorControls>
-			<div { ...blockProps }> { displayAuthor } </div>
+			<div { ...blockProps }>
+				{ supportsAuthor
+					? displayAuthor
+					: sprintf(
+							// translators: %s: Name of the post type e.g: "post".
+							__(
+								'This post type (%s) does not support the author'
```
Comment (Mamaduka) [inline packages/block-library/src/post-author-name/edit.js]:
```suggestion
								'This post type (%s) does not support the author.'
```

**Review by Mamaduka — APPROVED** [review]

**Pull request merged**